subsys/bluetooth/host/scan.c

```diff
@@ -639,22 +657,50 @@ void bt_hci_le_per_adv_report(struct net_buf *buf)
 	info.addr = &per_adv_sync->addr;
 	info.sid = per_adv_sync->sid;
 
-	if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL) {
-		/* Handling of incomplete reports is currently not
-		 * handled in the host. The remaining advertising
-		 * reports may therefore contain partial data.
-		 */
-		BT_WARN("Incomplete per adv report");
-	}
-
-	SYS_SLIST_FOR_EACH_CONTAINER(&pa_sync_cbs, listener, node) {
-		if (listener->recv) {
-			net_buf_simple_save(&buf->b, &state);
-
-			buf->len = evt->length;
-			listener->recv(per_adv_sync, &info, &buf->b);
-
-			net_buf_simple_restore(&buf->b, &state);
+	if (!per_adv_sync->report_truncated) {
+		if (per_adv_sync->reassembly == NULL) {
+			if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE) {
+				/* No reassembly buffer and a complete report.
+				 * Forward it without an extra copy.
+				 */
+				bt_hci_le_per_adv_report_recv(per_adv_sync, buf, &info);
+			} else {
+				BT_WARN("No reassembly buffer. Dropping the report");
+				per_adv_sync->report_truncated = false;
+			}
+		} else {
+			if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE &&
+			    per_adv_sync->reassembly->len == 0) {
+				/* We have not received any partial data before.
+				 * This buffer can be forwarded without an extra copy.
+				 */
+				bt_hci_le_per_adv_report_recv(per_adv_sync, buf, &info);
+				net_buf_reset(per_adv_sync->reassembly);
```

Comment:
Shouldn't we always reset the reassembly in case of a BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE regardless of whether `per_adv_sync->reassembly == NULL`? 

A periodic advertiser cannot send 2 SDUs interleaved, i.e. we cannot receive segments S1 and S2, and then receive a single complete PDU and then receive segment S3 afterwards.

Reply:
```net_buf_reset``` dereferences its argument, so it must not be called when ```per_adv_sync->reassembly == NULL```